### ISSUE TITLE
[ASNetworkImageNode] Suggestions / Help: Custom download and cache behaviour

### ISSUE THREAD

**Issue by flovouin** [facebook/AsyncDisplayKit#2629]
Hi guys,

I asked a question on Slack and got some great answers, but I'm still not entirely sure how to approach this problem.
I need to customize the cache and downloader for the `ASNetworkImageNode`. Basically I'd like to be able to do two things:

- For the downloader, "redirect" incoming requests to another URL. Please note that this is not a standard redirect, the destination URL is obtained asynchronously from another component inside the app. However the fetched images should be added to the cache with their original URL as key.
- For the cache, be able to manually add the image for a given key ("original" URL).

The reason for this is two fold (and matches the two bullets above):
- The original URLs set in `imageNode.URL` are not actual remote URLs. This follows a more global pattern in my app. Another service inside the app is in the charge of authentication / authorization, and asynchronously returns a temporary URL that can be used to download the resource. The temporary URL can't be used as a cache key, and I'd rather use the original URL, which is humanly readable, unique, and "immutable" for the resource.
- Because some of these images might be generated locally, remote requests aren't always necessary. I'd like to be able to add an image to the cache as soon as it's generated, independently from whether it's actually requested by an `ASNetworkImageNode`. I know the key / unique URL for the image, as it's shared across services in the app.

I took a look at the implementation of the ASPINRemoteImageDownloader. Although I can think of several ways I could achieve this, they are either based on assumptions on the internal representation of objects in the `PINCache`, or imply a non-negligible number of lines of code to write :D
My goal is obviously to keep as much of the default behaviour as possible, and that's why I'm turning to you guys for advice.

Any thoughts?

Cheers,

Flo

**Comment by appleguy** [comment]
@flovouin Hi!  Thanks for filing :)  I replied on Slack, but this is a good place for more information too.

> @flo I'd need to learn a bit more about what you're trying to do - may be helpful to hear concretely what you're doing to transform the URL to tell what could support it.  I'd recommend looking for -didEnterPreloadState (using 2.0-beta.1 or master) and at that time, generate the URL.  Basically use a custom subclass of ASNetworkImageNode, set your "sourceURL" on it, and then have the conversion to the "true" URL happen at a stage earlier — then set the actual .URL property to the correct URL.  If it isn't too expensive to do the conversions then definitely just do them at node init time instead, but doing them at Preload time will result in only the minimum number of URLs being converted (same number as will actually be fetched) so this should work

**Comment by flovouin** [comment]
@appleguy Thanks for the quick reply!
I set `imageNode.URL` to "fake", internal URLs e.g. `assets/thumbnails/1234`. A "service" in my app is in charge of converting these to actual remote URLs, by choosing which API end point to call with proper credentials e.g. `GET foo.com/thumbnails/1234 Authorization: 4321`. The remote end point then returns a temporary URL (a signed CloudFront URL).
I was hoping I could do all that without having to pass credentials around, hence by only specifying `assets/thumbnails/1234` to the image node (and then using a custom downloader).
If you find this workflow convoluted I couldn't really hold it against you :D The reason for this is partly based on the fact that I must do something similar for videos (using a `ResourceLoaderDelegate`), and that in this case I don't really have a choice. However I've found a solution that works great and clearly separates concerns (although my explanation of it might be a little foggy).

Now the second bullet of my original question is also something I'd really like to be able to do: manually add an image to the cache. Indeed, some of the images might be generated locally, and in this case I'd like to be able to set image in the cache for key `assets/thumbnails/1234` (again, preferably without having to specify actual remote end points). This would avoid an unnecessary remote request, that might fail if the image is still uploading or if the user has no connection at that time.

All of this boils down to two features:
- Being able to specify a downloader for which I could asynchronously redirect my fake URLs, while keeping its behaviour as close as possible to the default downloader when it comes to fetching the remote resource. (But the resource would have to be cached as the fake URL, not the temporary one.)
- Being able to manually add an image to the cache for a given fake URL / key.

By looking at the code of PINRemoteImage I've got a few ideas on how to achieve this, but it looks like I would have to rewrite a lot of error-prone code, that has already been battle tested in PINRemoteImage. I therefore posted this issue to ask you guys if you could think of a solution that would keep as much of the default behaviour as possible.

Cheers,

(P.S. I use the word "feature" but this is by no means a feature request. I just thought that given how long my post is simply to describe the issue, this would be a better fit for GitHub than Slack. :-))

**Comment by flovouin** [comment]
Okay so after looking at the code of the `PINRemoteImageManager`, here's what I did:

- Subclassed `PINRemoteImageManager` to:
  - Override `cacheKeyForURL`, which allows me to set the same key for "redirected URLs" and "fake URLs". This ensures that the downloader stores the image in the cache as the "fake URL" key.
  - Add a method to manually cache a given `UIImage` as a given "fake URL" key. Somehow accessing `materializeAndCacheObject` would probably be better (such that I wouldn't reimplement the caching logic), but unfortunately it's private.
- Created an object that implements both `ASImageCacheProtocol` and `ASImageDownloaderProtocol`, pretty much like `ASPINRemoteImageDownloader`. However because I have an extra step of asynchronous redirection before triggering the download, I have to issue my own identifiers first, and map them to the identifiers returned by the `PINRemoteImageManager` once I actually start the download. Although this works, this is a lot of error-prone code. I hope I got it right. ^^

As this solution does exactly what I wanted (although there might be a simpler approach), I'm closing the issue. ;-)

Thanks!
